staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go

```diff
@@ -122,18 +130,35 @@ func (c compiler) CompileCELExpression(expression string, envType environment.Ty
 		return resultError("unexpected compilation error: "+err.Error(), apiservercel.ErrorTypeInternal)
 	}
 	prog, err := env.Program(ast,
+		// cel.CostLimit is also set to the VAP PerCallLimit as part of
+		// the base environment.
+		//
+		// This call here should override that. In practice it shouldn't
+		// matter because the limits are the same.
+		cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost),
+		cel.CostTracking(estimator),
 		cel.InterruptCheckFrequency(celconfig.CheckFrequency),
 	)
 	if err != nil {
 		return resultError("program instantiation failed: "+err.Error(), apiservercel.ErrorTypeInternal)
 	}
-	return CompilationResult{
+
+	compilationResult := CompilationResult{
 		Program:     prog,
 		Expression:  expression,
 		OutputType:  ast.OutputType(),
 		Environment: env,
 		emptyMapVal: env.CELTypeAdapter().NativeToValue(map[string]any{}),
 	}
+
+	costEst, err := env.EstimateCost(ast, estimator)
+	if err != nil {
+		compilationResult.Error = &apiservercel.Error{Type: apiservercel.ErrorTypeInternal, Detail: "cost estimation failed: " + err.Error()}
+		return compilationResult
+	}
+
+	compilationResult.MaxCost = costEst.Max
```

Comment:
If the cost limit is a const, why don't check it here directly?

Reply:
I'm using the same approach as for VAP here: the validation code generates a validation error when the limit is exceeded.

This separates unrelated concerns:
- how to compile (here)
- what limits to impose (API)

Reply:
... although I must admit that the separation is not perfect because of the `cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost)` call. I'm still not sure whether such a runtime limit makes sense (https://github.com/kubernetes/kubernetes/pull/128101#discussion_r1803421102).

Reply:
[Removing the runtime cost checking](https://github.com/kubernetes/kubernetes/pull/128101#discussion_r1805053152) removed this usage of `resourceapi.CELSelectorExpressionMaxCost` in `compile.go`.

Reply:
... and now it's back as a default for some caller-specified `Options` struct.